synapse/handlers/_base.py

```diff
@@ -56,41 +85,39 @@ def _filter_events_for_client(self, user_id, events):
             )
         )
 
-        def allowed(event, state):
-            if event.type == EventTypes.RoomHistoryVisibility:
-                return True
+        events_to_return = []
+        events_to_return_if_later_joined = []
+        for event in events:
+            state = event_id_to_state[event.event_id]
 
-            membership_ev = state.get((EventTypes.Member, user_id), None)
-            if membership_ev:
-                membership = membership_ev.membership
+            membership_event = state.get((EventTypes.Member, user_id), None)
+            if membership_event:
+                membership = membership_event.membership
             else:
-                membership = Membership.LEAVE
+                membership = None
 
             if membership == Membership.JOIN:
-                return True
+                events_to_return.extend(events_to_return_if_later_joined)
+                events_to_return_if_later_joined = []
```

Comment:
Is it possible for an event to be marked as "include_if_later_joined" after an event with membership JOIN has been processed? Should the event be included in that case?